### PR TITLE
mime: compute full body md5

### PR DESCRIPTION
**Pull request by catenacyber**
Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
Preliminary work for https://redmine.openinfosecfoundation.org/issues/3487

Describe changes:
- computes full `email.body_md5`

Needed before #9036 

```
SV_BRANCH=pr/1252
```
https://github.com/OISF/suricata-verify/pull/1252

### CODE REVIEW

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 14717

File: src/util-decode-mime.c [hunk 1]
```diff
@@ -2348,6 +2337,18 @@ static int ProcessMimeEntity(const uint8_t *buf, uint32_t len,
                 MAX_LINE_LEN);
     }
 
+    if ((state->state_flag != HEADER_READY && state->state_flag != HEADER_STARTED) ||
```
Comment (jasonish) [inline src/util-decode-mime.c:2340]: More of a question.. Why not check state >= HEADER_STARTED?

**Comment by jasonish** [comment]
With respect to what is the full body MD5, I see the check for `CTNT_IS_BODYPART`.  Are we parsing the header, then taking the rest of the body verbatim for the MD5?  Or are there still parts after the header that won't be computed as part of the body?

File: src/util-decode-mime.c (hunk 1 above)
Comment (catenacyber) [reply src/util-decode-mime.c:2340]: Same condition (or rather contrary condition) is used below at line 2353 before calling `ProcessMimeHeaders` xor `ProcessMimeBody`

**Comment by catenacyber** [comment]
> Are we parsing the header, then taking the rest of the body verbatim for the MD5? 

That is what I would expect

> Or are there still parts after the header that won't be computed as part of the body?

Today, some headers/boundaries are not part of the body MD5

**Comment by jasonish** [comment]
> > Are we parsing the header, then taking the rest of the body verbatim for the MD5?
> 
> That is what I would expect

So that is what you think we should do, but are currently not?

**Comment by catenacyber** [comment]
> > > Are we parsing the header, then taking the rest of the body verbatim for the MD5?
> > 
> > 
> > That is what I would expect
> 
> So that is what you think we should do, but are currently not?

Exactly

The current logic is too complex to express about what is not taken in the md5

File: src/util-decode-mime.c [hunk 2]
```diff
@@ -2348,6 +2337,18 @@ static int ProcessMimeEntity(const uint8_t *buf, uint32_t len,
...
+            (state->stack->top->data->ctnt_flags & CTNT_IS_BODYPART)) {
+        if (!g_disable_hashing) {
```
[the remark this answers is not in the document]
Comment (catenacyber) [reply src/util-decode-mime.c:2342]: ok

**Comment by catenacyber** [comment]
Replaced by https://github.com/OISF/suricata/pull/9100

**Pull request closed without merging**